[jit] implement generic hashing operator

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44047 [jit] implement generic hashing operator**

We want to hash arbitrary ivalues by binding in our generic IValue
hasher into TorchScript.

closes #44038

Differential Revision: [D23480879](https://our.internmc.facebook.com/intern/diff/D23480879)

suo committed:
[jit] implement generic hashing operator

We want to hash arbitrary ivalues by binding in our generic IValue
hasher into TorchScript.

closes #44038

[ghstack-poisoned]

eellison commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -676,6 +676,12 @@ void hashValue(Stack* stack) {
 
+void hashGenericIValue(Stack* stack) {
+  auto value = pop(stack);
+  auto hash = (int64_t)IValue::HashAliasedIValue()(value);
```
This hash is only suppose to work for aliasing ivalues: https://github.com/pytorch/pytorch/blob/b44a10c1791f3f1d27171e6d012ca5f661e58aa9/aten/src/ATen/core/ivalue.h#L679
in python immutable scalars are hashed by value:
`hash((1, 2)) == hash((1, 2))` which this doesnt do

suo replied:
Ah, okay. So I guess I'll define a canonical `IValue::hash`

suo committed:
Update on "[jit] implement generic hashing operator"


We want to hash arbitrary ivalues by binding in our generic IValue
hasher into TorchScript.

closes #44038

Differential Revision: [D23480879](https://our.internmc.facebook.com/intern/diff/D23480879)

[ghstack-poisoned]
